Source/Core/Core/Debugger/BranchWatch.h

```diff
@@ -0,0 +1,286 @@
+// Copyright 2024 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#include <cstddef>
+#include <cstdio>
+#include <functional>
+#include <type_traits>
+#include <unordered_map>
+#include <vector>
+
+#include "Common/BitUtils.h"
+#include "Common/CommonTypes.h"
+#include "Core/PowerPC/Gekko.h"
+
+namespace Core
+{
+class CPUThreadGuard;
+}
+
+namespace Core
+{
+struct FakeBranchWatchCollectionKey
+{
+  u32 origin_addr;
+  u32 destin_addr;
+};
+struct BranchWatchCollectionKey : FakeBranchWatchCollectionKey
+{
+  UGeckoInstruction original_inst;
+};
+struct BranchWatchCollectionValue
+{
+  std::size_t total_hits = 0;
+  std::size_t hits_snapshot = 0;
+};
+}  // namespace Core
+
+template <>
+struct std::hash<Core::BranchWatchCollectionKey>
+{
+  std::size_t operator()(const Core::BranchWatchCollectionKey& s) const noexcept
+  {
+    return std::hash<u64>{}(
+        Common::BitCast<u64>(static_cast<const Core::FakeBranchWatchCollectionKey&>(s)));
+  }
+};
+
+namespace Core
+{
+inline bool operator==(const BranchWatchCollectionKey& lhs,
+                       const BranchWatchCollectionKey& rhs) noexcept
+{
+  std::hash<BranchWatchCollectionKey> hash;
+  return hash(lhs) == hash(rhs) && lhs.original_inst.hex == rhs.original_inst.hex;
+}
+
+enum class BranchWatchSelectionInspection : unsigned char
+{
+  SetOriginNOP = 1u << 0,
+  SetDestinBLR = 1u << 1,
+  SetOriginSymbolBLR = 1u << 2,
+  SetDestinSymbolBLR = 1u << 3,
+  EndOfEnumeration,
+};
+
+constexpr BranchWatchSelectionInspection operator|(BranchWatchSelectionInspection lhs,
+                                                   BranchWatchSelectionInspection rhs)
+{
+  using underlying_t = std::underlying_type_t<BranchWatchSelectionInspection>;
+  return static_cast<BranchWatchSelectionInspection>(static_cast<underlying_t>(lhs) |
+                                                     static_cast<underlying_t>(rhs));
+}
+
+constexpr BranchWatchSelectionInspection operator&(BranchWatchSelectionInspection lhs,
+                                                   BranchWatchSelectionInspection rhs)
+{
+  using underlying_t = std::underlying_type_t<BranchWatchSelectionInspection>;
+  return static_cast<BranchWatchSelectionInspection>(static_cast<underlying_t>(lhs) &
+                                                     static_cast<underlying_t>(rhs));
+}
+
+constexpr BranchWatchSelectionInspection& operator|=(BranchWatchSelectionInspection& self,
+                                                     BranchWatchSelectionInspection other)
+{
+  return self = self | other;
+}
+
+class BranchWatchCollection final
+    : public std::unordered_map<BranchWatchCollectionKey, BranchWatchCollectionValue>
```

Comment:
Subclassing STL types like this feels really icky... Is there an advantage to doing this over just `using BranchWatchCollection = std::unordered_map<BranchWatchCollectionKey, BranchWatchCollectionValue>`?

Reply:
You can't forward declare a using declaration like you can a class / struct. I think in an earlier iteration of the design, I needed to forward declare some class from BranchWatch.h in another header to skip the include, and switched all of the using declarations over to subclassing to be consistent and future-proof. In my opinion, I prefer this way, but am not overly attached to it if it better fits the style to use using declarations.